crates/sim2h/src/cache.rs

```diff
@@ -178,13 +184,29 @@ impl Space {
             .filtered_by_entry_hash(|entry_hash| {
                 naive_sharding_should_store(
                     agent_loc,
-                    entry_location(&self.crypto, &entry_hash),
+                    self.entry_location(entry_hash),
                     self.agents.len() as u64,
                     redundant_count,
                 )
             })
     }
 
+    pub fn entry_location(&self, entry_hash: &EntryHash) -> Location {
+        let cached_location: Option<Location> = self
+            .entry_locations
+            .read()
+            .ok()
+            .and_then(|map| map.get(entry_hash).cloned());
```

Comment:
prefer

- https://doc.rust-lang.org/std/collections/struct.HashMap.html#method.entry
- https://doc.rust-lang.org/std/collections/hash_map/enum.Entry.html#method.or_insert_with

then it only has to hash/probe the hashmap once